File: Data Structures & Algorithms/binary-tree-right-side-view/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if(root == nullptr)
            return {};

        std::vector<int> res;

        std::queue<TreeNode*> rootQueue;
        rootQueue.push(root);

        while(rootQueue.size() > 0) {
            int size = rootQueue.size();

            for(int i = 0; i < size; i++) {
                TreeNode* front = rootQueue.front();
                rootQueue.pop();

                if(i == size - 1) {
                    res.emplace_back(front->val);
                }

                if(front->left != nullptr) 
                    rootQueue.push(front->left);

                if(front->right != nullptr)
                    rootQueue.push(front->right);
            }
        }

        return res; 
    }
};

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-3.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        if(root == null)
            return listOf()

        var list : MutableList<List<Int>> = mutableListOf()

        val queue : ArrayDeque<TreeNode> = ArrayDeque()
        queue.add(root)

        while(queue.size > 0) {
            var currentList = mutableListOf<Int>()

            for(i in queue.indices) {
                val front = queue.first()
                queue.removeFirst()

                currentList.add(front.`val`)

                if(front.left != null) {
                    queue.add(front.left)
                }

                if(front.right != null) {
                    queue.add(front.right)
                }
            }

            list.add(currentList.toList())
        }

        return list 
    }
}
